File: json.go
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"strconv"
	"strings"
)

func generateJson(inputMap map[string]any, envMap map[string]string, prefix string) {
	prefix = strings.TrimSpace(prefix)
	if len(prefix) > 0 {
		prefix = fmt.Sprintf("%s_", prefix)
	}

	for k, v := range inputMap {
		name := fmt.Sprintf("%s%s", prefix, k)

		switch v := v.(type) {
		case string:
			envMap[name] = v
		case float64:
			if v == math.Trunc(v) {
				envMap[name] = fmt.Sprintf("%d", int(v))
			} else {
				envMap[name] = strconv.FormatFloat(v, 'f', -1, 64)
			}
		case map[string]any:
			generateJson(v, envMap, name)
		case []any:
			panic(fmt.Sprintf("key '%s': arrays are not supported", name))
		default:
			panic(fmt.Sprintf("key: '%s': unknown type", name))
		}
	}
}

func GenerateJson(rawJson string) (map[string]string, error) {
	envMap := make(map[string]string)

	var rawJsonMap map[string]any
	if err := json.Unmarshal([]byte(rawJson), &rawJsonMap); err != nil {
		return nil, fmt.Errorf("failed to unmarshal input json: %s", err)
	}

	generateJson(rawJsonMap, envMap, "")

	return envMap, nil
}

File: src/json.js
import { InvalidContentError, UnsupportedContentError } from './errors.js';

function generateJson(inputObj, prefix) {
    let envMap = new Map();

	if (prefix.length > 0) {
		prefix = prefix + "_";
	}

    const m = new Map(Object.entries(inputObj));

    for (const [k, v] of m) {
        const name = prefix + k;

        if (v instanceof Array) {
            throw new UnsupportedContentError('key \'' + name + '\': arrays are not supported')
        } else if (typeof v == "string") {
            envMap.set(name,  v);
        } else if (typeof v == "number") {
            envMap.set(name,  "" + v);
        } else if (typeof v == "object") {
            envMap = new Map([...envMap, ...generateJson(v, name)]);
        }
    }

    return envMap;
}

export default function GenerateJson(rawJson, prefix)  {
    let rawJsonObject = Object;

    try {
        rawJsonObject = JSON.parse(rawJson);
    } catch(error) {
        throw new InvalidContentError('failed to parse JSON: ' + error);
    }

    return generateJson(rawJsonObject, prefix);
}
